### PR TITLE
[FLINK-23040][table-common] Consider ConfigOption fallback keys in FactoryUtil

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

Adds support for fallback keys (incl. deprecated keys) to FactoryUtil. Deprecated keys are highlighted in error messages.

## Brief change log

- Simplify util by abstracting common helper code in a super class.
- Add support for fallback keys in validation.

## Verifying this change

This change added tests and can be verified as follows: `FactoryUtilTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java [hunk 1]
```diff
@@ -578,15 +621,27 @@ public void validateExcept(String... prefixesToSkip) {
...
+            return allOptions;
+        }
+    }
+
+    /**
+     * Helper utility for validating all options for a {@link CatalogFactory}.
+     *
+     * @see #createCatalogFactoryHelper(CatalogFactory, CatalogFactory.Context)
+     */
+    public static class CatalogFactoryHelper extends FactoryHelper<CatalogFactory> {
```
Comment (Airblader) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java:641]: `@PublicEvolving`?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java [hunk 2]
```diff
@@ -596,32 +651,18 @@ public ReadableConfig getOptions() {
      *
      * @see #createTableFactoryHelper(DynamicTableFactory, DynamicTableFactory.Context)
      */
-    public static class TableFactoryHelper {
-
-        private final DynamicTableFactory tableFactory;
+    public static class TableFactoryHelper extends FactoryHelper<DynamicTableFactory> {
```
Comment (Airblader) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java:654]: Unlike `CatalogFactoryHelper`, this wasn't `@PublicEvolving` before, but it seems like it should be as well.

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/factories/TestDynamicTableFactory.java [hunk 3]
```diff
@@ -50,10 +50,16 @@
     public static final String IDENTIFIER = "test-connector";
 
     public static final ConfigOption<String> TARGET =
-            ConfigOptions.key("target").stringType().noDefaultValue();
+            ConfigOptions.key("target")
+                    .stringType()
+                    .noDefaultValue()
+                    .withDeprecatedKeys("deprecated-target");
```
Comment (Airblader) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/factories/TestDynamicTableFactory.java:56]: (Going on a bit of a tangent here, but we should eventually get to a point where we can test these things in isolation rather than having a single factory which tries to implement every test(able) scenario at once. Of course this is out of scope for this PR.)

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java:641]: The upper class is already declared `@PublicEvolving`.

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/factories/TestDynamicTableFactory.java (hunk 3 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/factories/TestDynamicTableFactory.java:56]: I agree. It should be easier to hook into the discovery process.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java (hunk 1 above)
Comment (Airblader) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java:641]: Ah, I guess that makes sense. I went by the fact that it was annotated previously already – which was done by me. :-)

**Review by Airblader — APPROVED** [review]
LGTM
